### PR TITLE
YARN-6668

### PR DESCRIPTION
**Pull request by szegedim**
YARN-6668

### CODE REVIEW

File: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/linux/resources/CGroupsResourceCalculator.java [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * A cgroups file-system based Resource calculator without the process tree
+ * features.
+ */
+
+public class CGroupsResourceCalculator extends ResourceCalculatorProcessTree {
```
Comment (haibchen) [inline hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/linux/resources/CGroupsResourceCalculator.java]: This is pretty neat! :-) I have not looked closely at the details of this class. One reminder is we want to use this as well to get the aggregated container utilization. Now, we just do accumulation while iterating over all container process (see ContainerMonitorImpl.setContainersUtilization()) in container monitor, which can be stale if there are a lot of containers running.

File: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/linux/resources/CGroupsResourceCalculator.java [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+  public float getCpuUsagePercent() {
+    readTotalProcessJiffies();
+    cpuTimeTracker.updateElapsedJiffies(
+        processTotalJiffies,
+        clock.getTime());
+    return cpuTimeTracker.getCpuTrackerUsagePercent();
+  }
+
+  @Override
+  public long getCumulativeCpuTime() {
```
Comment (haibchen) [inline hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/linux/resources/CGroupsResourceCalculator.java]: Why not use cpuacct.usage?

File: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/linux/resources/CGroupsResourceCalculator.java (hunk 1 above)
Comment (szegedim) [inline hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/linux/resources/CGroupsResourceCalculator.java]: It supports aggregated container utilization by giving a null pid

File: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/monitor/ContainersMonitorImpl.java [hunk 3]
```diff
@@ -588,6 +589,31 @@ private void initializeProcessTrees(
...
+     * Get the best process tree calculator.
+     * @param pId container process id
+     * @return process tree calculator
+     */
+    private ResourceCalculatorProcessTree
+        getResourceCalculatorProcessTree(String pId) {
+      ResourceCalculatorProcessTree pt = null;
+
+      // CGroups is best in perforance, so try to use it, if it is enabled
+      if (processTreeClass == null &&
```
Comment (szegedim) [inline hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/monitor/ContainersMonitorImpl.java]: Impossible. CGroupsResourceCalculator relies on cgroups that are present only in node manager. ResourceCalculatorProcessTree is hadoop-yarn.common package

File: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/linux/resources/CGroupsResourceCalculator.java (hunk 2 above)
Comment (szegedim) [inline hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/linux/resources/CGroupsResourceCalculator.java]: Compatibility. The original CPU time from procfs was based on /proc/<>/stat which are the same counters user+kernel in jiffies.

**Pull request closed without merging**